File: openoversea/src/main/java/com/bytedance/sdk/open/tiktok/share/Share.java
package com.bytedance.sdk.open.tiktok.share;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;

import com.bytedance.sdk.open.tiktok.base.Anchor;
import com.bytedance.sdk.open.tiktok.base.MediaContent;
import com.bytedance.sdk.open.tiktok.common.constants.Constants;
import com.bytedance.sdk.open.tiktok.base.MicroAppInfo;
import com.bytedance.sdk.open.tiktok.common.constants.Keys;
import com.bytedance.sdk.open.tiktok.common.model.BaseReq;
import com.bytedance.sdk.open.tiktok.common.model.BaseResp;

import java.util.ArrayList;
import java.util.HashMap;

// TODO: chen.wu change it when auth response is migrated and base is replaced with Base in response
public class Share {

    private static final String TAG = "Aweme.OpenSDK.Share";
    public static final int VIDEO = 0;
    public static final int IMAGE = 1;
    public enum Format {
        DEFAULT(0),
        GREEN_SCREEN(1);
        private final int value;

        Format(int i) {
            this.value = i;
        }
        public int getValue() {
            return value;
        }
    }
    public static class Request extends BaseReq {

        public int mTargetSceneType = 0;

        public ArrayList<String> mHashTagList;
        public Format mShareFormat;
        public MediaContent mMediaContent;
        public MicroAppInfo mMicroAppInfo;

//        public AnchorObject mAnchorInfo;
        public Anchor mAnchorInfo;
        public String mCallerPackage;

        public String mClientKey;

        public String mState;

        public String mExtra;
        public String mAnchorSourceType;

        // Allow developers to include extra share options such as enable/disable certain actions
        // in the creation flow
        public HashMap<String, Integer> extraShareOptions;

        public Request() {
            super();
        }

        public Request(Bundle bundle) {
            fromBundle(bundle);
        }

        @Override
        public int getType() {
            return Constants.TIKTOK.SHARE_REQUEST;
        }

        @SuppressLint("MissingSuperCall")
        @Override
        public void fromBundle(Bundle bundle) {
            super.fromBundle(bundle);
            this.mCallerPackage = bundle.getString(Keys.Share.CALLER_PKG);
            this.callerLocalEntry = bundle.getString(Keys.Share.CALLER_LOCAL_ENTRY);
            this.mState = bundle.getString(Keys.Share.STATE);
            this.mClientKey = bundle.getString(Keys.Share.CLIENT_KEY);
            this.mTargetSceneType =
                    bundle.getInt(Keys.Share.SHARE_TARGET_SCENE, Keys.Scene.LANDPAGE_SCENE_DEFAULT);
            this.mHashTagList = bundle.getStringArrayList(Keys.Share.SHARE_HASHTAG_LIST);
            this.mMediaContent = MediaContent.Companion.INSTANCE.fromBundle(bundle); // MediaContent.Builder.fromBundle(bundle);
            this.mMicroAppInfo = MicroAppInfo.unserialize(bundle);
            this.mAnchorInfo = Anchor.Companion.INSTANCE.fromBundle(bundle);
            int shareFormatValue = bundle.getInt(Keys.Share.SHARE_FORMAT);
            switch (shareFormatValue) {
                case 1: {
                    this.mShareFormat = Format.GREEN_SCREEN;
                } break;
                default:
                    this.mShareFormat = Format.DEFAULT;
            }
        }

        @SuppressLint("MissingSuperCall")
        @Override
        public void toBundle(Bundle bundle) {
            super.toBundle(bundle);
            bundle.putString(Keys.Share.CALLER_LOCAL_ENTRY, callerLocalEntry);
            bundle.putString(Keys.Share.CLIENT_KEY, mClientKey);
            bundle.putString(Keys.Share.CALLER_PKG, mCallerPackage);
            bundle.putString(Keys.Share.STATE, mState);

            bundle.putAll(this.mMediaContent.toBundle());
            bundle.putInt(Keys.Share.SHARE_TARGET_SCENE, mTargetSceneType);
            if (mHashTagList != null && mHashTagList.size() > 0) {
                bundle.putString(Keys.Share.SHARE_DEFAULT_HASHTAG, mHashTagList.get(0)); // 兼容旧版本aweme
                bundle.putStringArrayList(Keys.Share.SHARE_HASHTAG_LIST, mHashTagList);
            }

            // 670 add micro app
            if (mMicroAppInfo != null) {
                mMicroAppInfo.serialize(bundle);
            }
            // 920 add anchor
            if (mAnchorInfo != null) {
                if (mAnchorInfo.getAnchorBusinessType() == 10){
                    bundle.putAll(mAnchorInfo.toBundle());
                }
            }
            bundle.putInt(Keys.Share.SHARE_FORMAT, mShareFormat.value);
        }


        @SuppressLint("MissingSuperCall")
        public boolean checkArgs() {
            if (this.mMediaContent == null) {
                Log.e(TAG, "checkArgs fail ,mediaContent is null");
                return false;
            } else {
//                return this.mMediaContent.checkArgs();
                return true;
            }
        }
    }


    public static class Response extends BaseResp {
        public String state;

        public int subErrorCode;


        public Response() {
        }

        public Response(Bundle bundle) {
            fromBundle(bundle);
        }

        @Override
        public int getType() {
            return Constants.TIKTOK.SHARE_RESPONSE;
        }

        @SuppressLint("MissingSuperCall")
        @Override
        public void fromBundle(Bundle bundle) {
            this.errorCode = bundle.getInt(Keys.Share.ERROR_CODE);
            this.errorMsg = bundle.getString(Keys.Base.ERROR_MSG);
            this.extras = bundle.getBundle(Keys.Base.EXTRA); // EXTRAS 复用老base
            this.state = bundle.getString(Keys.Share.STATE);
            this.subErrorCode = bundle.getInt(Keys.Share.SHARE_SUB_ERROR_CODE, -1000); // set default value

        }

        @SuppressLint("MissingSuperCall")
        @Override
        public void toBundle(Bundle bundle) {
            bundle.putInt(Keys.Share.ERROR_CODE, errorCode);
            bundle.putString(Keys.Base.ERROR_MSG, errorMsg);
            bundle.putInt(Keys.Share.TYPE, getType());
            bundle.putBundle(Keys.Base.EXTRA, extras); // EXTRAS 复用老base
            bundle.putString(Keys.Share.STATE, state);
            bundle.putInt(Keys.Share.SHARE_SUB_ERROR_CODE, subErrorCode);

        }
    }
}

File: openoversea/src/main/java/com/bytedance/sdk/open/tiktok/share/Share.kt
package com.bytedance.sdk.open.tiktok.share

/*
    Copyright 2022 TikTok Pte. Ltd.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

import android.os.Bundle
import com.bytedance.sdk.open.tiktok.base.Anchor
import com.bytedance.sdk.open.tiktok.base.MediaContent
import com.bytedance.sdk.open.tiktok.common.constants.Constants
import com.bytedance.sdk.open.tiktok.common.constants.Keys
import com.bytedance.sdk.open.tiktok.common.model.Base
import com.bytedance.sdk.open.tiktok.common.model.ResultActivityComponent
import com.bytedance.sdk.open.tiktok.utils.AppUtils

class Share {
    enum class Format(val format: Int) {
        DEFAULT(0),
        GREEN_SCREEN(1);
    }

    enum class MediaType {
        VIDEO,
        IMAGE;
    }

    data class Request(
        val mediaContent: MediaContent,
        val targetSceneType: Int = 0,
        val hashTagList: ArrayList<String> = arrayListOf(),
        val shareFormat: Format = Format.DEFAULT,
        val anchor: Anchor? = null,
        val state: String? = null,
        val shareExtra: String? = null,
        val extraShareOptions: HashMap<String, Any>? = null,
        override val resultActivityComponent: ResultActivityComponent,
    ) : Base.Request() {

        override var type: Int = Constants.TIKTOK.SHARE_REQUEST

        override fun validate(): Boolean {
            return mediaContent.validate()
        }

        override fun toBundle(clientKey: String): Bundle {
            return super.toBundle().apply {
                putString(Keys.Share.CLIENT_KEY, clientKey)
                putString(Keys.Share.CALLER_SDK_VERSION, Keys.VERSION)
                putString(Keys.Share.STATE, state)
                putAll(mediaContent.toBundle())
                putInt(Keys.Share.SHARE_FORMAT, shareFormat.format)
                putInt(Keys.Share.SHARE_TARGET_SCENE, targetSceneType)
                putString(Keys.Share.OPEN_PLATFORM_EXTRA, shareExtra)
                putSerializable(Keys.Share.EXTRA_SHARE_OPTIONS, extraShareOptions)

                hashTagList.let {
                    if (it.isNotEmpty()) {
                        putString(Keys.Share.SHARE_DEFAULT_HASHTAG, it[0])
                    }
                    putStringArrayList(Keys.Share.SHARE_HASHTAG_LIST, it)
                }
                anchor?.apply {
                    if (this.anchorBusinessType == 10) { // TODO: chen.wu check this anchor business type
                        putAll(this.toBundle())
                    }
                    putString(Keys.Share.ANCHOR_SOURCE_TYPE, sourceType)
                }

                putString(Keys.Share.CALLER_PKG, resultActivityComponent.packageName)
                putString(
                    Keys.Share.CALLER_LOCAL_ENTRY,
                    AppUtils.componentClassName(
                        packageName = resultActivityComponent.packageName,
                        classPath = resultActivityComponent.className
                    )
                )
            }
        }
    }

    data class Response(
        var state: String?,
        var subErrorCode: Int?,
        override val errorCode: Int,
        override val errorMsg: String?,
        override val extras: Bundle? = null,
    ) : Base.Response() {
        override val type: Int = Constants.TIKTOK.SHARE_RESPONSE
    }
}

internal fun Bundle.toShareResponse(): Share.Response {
    val state = getString(Keys.Share.STATE)
    val subErrorCode = getInt(Keys.Share.SHARE_SUB_ERROR_CODE)
    val errorCode = getInt(Keys.Share.ERROR_CODE)
    val errorMsg = getString(Keys.Share.ERROR_MSG)
    val extras = getBundle(Keys.Base.EXTRA)
    return Share.Response(
        state = state,
        subErrorCode = subErrorCode,
        errorCode = errorCode,
        errorMsg = errorMsg,
        extras = extras
    )
}
